File: cli/src/ssh-agent.rs
use base64::Engine;
use serde::Deserialize;
use ssh_agent_lib::agent::{listen, Session};
use ssh_agent_lib::error::AgentError;
use ssh_agent_lib::proto::{Identity, PublicCredential, SignRequest};
use ssh_agent_lib::ssh_key::{PublicKey, Signature};
use std::sync::Arc;
use tokio::net::UnixListener;
use tokio::sync::Mutex;
use phone_key_cli::{bluetooth, config};

const SOCKET_PATH: &str = "/tmp/phone-key-ssh-agent.sock";

#[derive(Deserialize)]

struct IdentityResponse {
    alias: String,
    #[serde(rename = "publicKeyBase64")]
    public_key_base64: String,
}

#[derive(Clone)]
struct AppSession {
    device_address: String,
    connection: Arc<Mutex<Option<bluetooth::BluetoothConnection>>>,
    identities: Vec<(PublicCredential, String)>,
}

impl AppSession {
    fn new(config: &config::Config) -> Self {
        let device_address = config
            .device_address
            .clone()
            .expect("device_address not set in ~/.phone-key.json");
        Self {
            device_address,
            connection: Arc::new(Mutex::new(None)),
            identities: Vec::new(),
        }
    }

    async fn send_message(&self, json: &str) -> Result<Vec<u8>, AgentError> {
        let mut guard = self.connection.lock().await;
        if guard.is_none() {
            *guard = Some(
                bluetooth::BluetoothConnection::connect(&self.device_address)
                    .await
                    .map_err(|e| AgentError::other(std::io::Error::new(std::io::ErrorKind::Other, e.to_string())))?,
            );
        }

        match guard
            .as_mut()
            .expect("connection just initialized")
            .send_message(json)
            .await
        {
            Ok(response) => Ok(response),
            Err(e) => {
                // Dispose broken connection so next call creates a fresh one
                *guard = None;
                eprintln!("[ssh-agent] Bluetooth connection error: {e}.");
                Err(AgentError::other(std::io::Error::new(
                    std::io::ErrorKind::Other,
                    e.to_string(),
                )))
            }
        }
    }
}

#[ssh_agent_lib::async_trait]
impl Session for AppSession {
    async fn request_identities(&mut self) -> Result<Vec<Identity>, AgentError> {
        let response = self
            .send_message(r#"{"type":"ssh-request-identities"}"#)
            .await?;

        let parsed: Vec<IdentityResponse> = serde_json::from_slice(&response).map_err(|e| {
            AgentError::other(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("failed to parse identities response: {e}"),
            ))
        })?;

        self.identities.clear();
        let mut identities = Vec::new();
        for item in parsed {
            let public_key_bytes = base64::engine::general_purpose::STANDARD
                .decode(&item.public_key_base64)
                .map_err(|e| {
                    AgentError::other(std::io::Error::new(
                        std::io::ErrorKind::Other,
                        format!("failed to base64-decode public key for '{}': {e}", item.alias),
                    ))
                })?;
            let public_key_str = String::from_utf8(public_key_bytes).map_err(|e| {
                AgentError::other(std::io::Error::new(
                    std::io::ErrorKind::Other,
                    format!("invalid UTF-8 in public key for '{}': {e}", item.alias),
                ))
            })?;
            eprintln!("{}", public_key_str);
            let pk = PublicKey::from_openssh(&public_key_str).map_err(|e| {
                AgentError::other(std::io::Error::new(
                    std::io::ErrorKind::Other,
                    format!("invalid public key for '{}': {e}", item.alias),
                ))
            })?;
            let credential = PublicCredential::Key(pk.key_data().clone());
            identities.push(Identity {
                credential: credential.clone(),
                comment: item.alias.clone(),
            });
            self.identities.push((credential, item.alias));
        }

        Ok(identities)
    }

    async fn sign(&mut self, request: SignRequest) -> Result<Signature, AgentError> {
        let alias = self
            .identities
            .iter()
            .find(|(cred, _)| cred == &request.credential)
            .map(|(_, alias)| alias.as_str())
            .ok_or_else(|| {
                AgentError::other(std::io::Error::new(
                    std::io::ErrorKind::NotFound,
                    "key not found in cached identities",
                ))
            })?;

        let data_b64 = base64::engine::general_purpose::STANDARD.encode(&request.data);
        let json = serde_json::json!({
            "type": "ssh-sign",
            "keyAlias": alias,
            "data": data_b64,
        });

        let response = self.send_message(&json.to_string()).await?;
        if response.is_empty() {
            return Err(AgentError::other(std::io::Error::new(
                std::io::ErrorKind::Other,
                "signing failed on device",
            )));
        }

        let algorithm = request.credential.key_data().algorithm();
        Signature::new(algorithm, response).map_err(|e| {
            AgentError::other(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("failed to construct signature: {e}"),
            ))
        })
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config = config::Config::load()?;
    let session = AppSession::new(&config);

    let _ = std::fs::remove_file(SOCKET_PATH);

    let listener = UnixListener::bind(SOCKET_PATH)?;
    println!("SSH agent listening on {SOCKET_PATH}");
    println!("export SSH_AUTH_SOCK={SOCKET_PATH}");

    listen(listener, session).await?;

    Ok(())
}

File: cli-go/src/phonekey/ssh-agent.go
package phonekey

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"os/signal"
	"sync"
	"syscall"

	"golang.org/x/crypto/ssh"
	"golang.org/x/crypto/ssh/agent"
)

const SocketPath = "/tmp/phone-key-ssh-agent.sock"

type identityResponse struct {
	Alias           string `json:"alias"`
	PublicKeyBase64 string `json:"publicKeyBase64"`
}

type requestIdentities struct {
	Type string `json:"type"`
}

type signRequest struct {
	Type     string `json:"type"`
	KeyAlias string `json:"keyAlias"`
	Data     string `json:"data"`
}

type AppAgent struct {
	mu         sync.Mutex
	conn       *UsbConnection
	identities []identityEntry
}

type identityEntry struct {
	key     *agent.Key
	cred    ssh.PublicKey
	comment string
}

func NewAppAgent() *AppAgent {
	return &AppAgent{}
}

func (a *AppAgent) ensureConnection() error {
	if a.conn != nil {
		return nil
	}
	conn, err := ConnectUsb()
	if err != nil {
		return err
	}
	a.conn = conn
	return nil
}

func (a *AppAgent) Disconnect() {
	if a.conn != nil {
		a.conn.Close()
		a.conn = nil
	}
}

func (a *AppAgent) sendAndReceive(msg interface{}) ([]byte, error) {
	payload, err := json.Marshal(msg)
	if err != nil {
		return nil, fmt.Errorf("marshal: %w", err)
	}
	response, err := a.conn.SendMessage(payload)
	if err != nil {
		a.Disconnect()
		return nil, err
	}
	return response, nil
}

func (a *AppAgent) List() ([]*agent.Key, error) {
	a.mu.Lock()
	defer a.mu.Unlock()

	if err := a.ensureConnection(); err != nil {
		return nil, err
	}

	resp, err := a.sendAndReceive(requestIdentities{Type: "ssh-request-identities"})
	if err != nil {
		return nil, err
	}

	var parsed []identityResponse
	if err := json.Unmarshal(resp, &parsed); err != nil {
		return nil, fmt.Errorf("parse identities: %w", err)
	}

	a.identities = nil
	var keys []*agent.Key
	for _, item := range parsed {
		pubKeyBytes, err := base64.StdEncoding.DecodeString(item.PublicKeyBase64)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[ssh-agent] base64 decode failed for '%s': %v\n", item.Alias, err)
			continue
		}

		pubKey, _, _, _, err := ssh.ParseAuthorizedKey(pubKeyBytes)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[ssh-agent] parse public key failed for '%s': %v\n", item.Alias, err)
			continue
		}

		key := &agent.Key{
			Format:  pubKey.Type(),
			Blob:    pubKey.Marshal(),
			Comment: item.Alias,
		}
		keys = append(keys, key)
		a.identities = append(a.identities, identityEntry{
			key:     key,
			cred:    pubKey,
			comment: item.Alias,
		})
	}

	return keys, nil
}

func (a *AppAgent) Sign(key ssh.PublicKey, data []byte) (*ssh.Signature, error) {
	a.mu.Lock()
	defer a.mu.Unlock()

	var entry *identityEntry
	for i := range a.identities {
		if a.identities[i].cred.Type() == key.Type() &&
			string(a.identities[i].cred.Marshal()) == string(key.Marshal()) {
			entry = &a.identities[i]
			break
		}
	}
	if entry == nil {
		return nil, fmt.Errorf("key not found in cached identities")
	}

	if err := a.ensureConnection(); err != nil {
		return nil, err
	}

	dataB64 := base64.StdEncoding.EncodeToString(data)
	resp, err := a.sendAndReceive(signRequest{
		Type:     "ssh-sign",
		KeyAlias: entry.comment,
		Data:     dataB64,
	})
	if err != nil {
		return nil, err
	}

	if len(resp) == 0 {
		return nil, fmt.Errorf("signing failed on device")
	}

	sig := &ssh.Signature{
		Format: key.Type(),
		Blob:   resp,
	}
	return sig, nil
}

func (a *AppAgent) Signers() ([]ssh.Signer, error) {
	return nil, nil
}

func (a *AppAgent) Add(key agent.AddedKey) error {
	return fmt.Errorf("adding keys is not supported")
}

func (a *AppAgent) Remove(key ssh.PublicKey) error {
	return fmt.Errorf("removing keys is not supported")
}

func (a *AppAgent) RemoveAll() error {
	return fmt.Errorf("removing keys is not supported")
}

func (a *AppAgent) Lock(passphrase []byte) error {
	return fmt.Errorf("locking is not supported")
}

func (a *AppAgent) Unlock(passphrase []byte) error {
	return fmt.Errorf("unlocking is not supported")
}

func RunSSHAgent() error {
	_ = os.Remove(SocketPath)

	listener, err := net.Listen("unix", SocketPath)
	if err != nil {
		return fmt.Errorf("listen on %s: %w", SocketPath, err)
	}
	defer os.Remove(SocketPath)

	fmt.Printf("SSH agent listening on %s\n", SocketPath)
	fmt.Printf("export SSH_AUTH_SOCK=%s\n", SocketPath)

	app := NewAppAgent()

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigCh
		fmt.Fprintln(os.Stderr, "\n[ssh-agent] shutting down...")
		app.Disconnect()
		listener.Close()
		os.Remove(SocketPath)
		os.Exit(0)
	}()

	for {
		conn, err := listener.Accept()
		if err != nil {
			select {
			case <-sigCh:
				return nil
			default:
				return fmt.Errorf("accept: %w", err)
			}
		}
		go func() {
			if err := agent.ServeAgent(app, conn); err != nil {
				fmt.Fprintf(os.Stderr, "[ssh-agent] session error: %v\n", err)
			}
		}()
	}
}
